### PR TITLE
Refactor: Cleanup coordinator duties for metadata cleanup

### PR DESCRIPTION
**Pull request by kfaraz**
### Changes
- Add an abstract class `MetadataCleanupDuty` which contains common code from the metadata cleanup duties
- Use `Duration` and `DateTime` objects to track `retainDuration`, `lastCleanupTime` and `cleanupPeriod` for better readability and cleaner log and error messages.
- Cleanup tests
- No functional change

### Classes to review
- `MetadataCleanupDuty`
- `KillCompactionConfig`
- `KillSupervisors`
- `KillSupervisorsCustomDuty`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/duty/KillCompactionConfig.java [hunk 1]
```diff
@@ -72,99 +66,85 @@ public KillCompactionConfig(
...
-      try {
-        RetryUtils.retry(
-            () -> {
-              final byte[] currentBytes = CoordinatorCompactionConfig.getConfigInByteFromDb(connector, connectorConfig);
-              final CoordinatorCompactionConfig current = CoordinatorCompactionConfig.convertByteToConfig(jacksonConfigManager, currentBytes);
-              // If current compaction config is empty then there is nothing to do
-              if (CoordinatorCompactionConfig.empty().equals(current)) {
-                log.info(
-                    "Finished running KillCompactionConfig duty. Nothing to do as compaction config is already empty.");
-                emitMetric(params.getEmitter(), 0);
```
Comment (findingrish) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/KillCompactionConfig.java:105]: are we emitting this metric?

File: server/src/main/java/org/apache/druid/server/coordinator/duty/KillCompactionConfig.java [hunk 2]
```diff
@@ -72,99 +66,85 @@ public KillCompactionConfig(
...
-                );
-                emitMetric(params.getEmitter(), compactionConfigRemoved);
-              } else if (result.isRetryable()) {
-                // Failed but is retryable
-                log.debug("Retrying KillCompactionConfig duty");
-                throw new RetryableException(result.getException());
-              } else {
-                // Failed and not retryable
-                log.error(result.getException(), "Failed to kill compaction configurations");
-                emitMetric(params.getEmitter(), 0);
```
Comment (findingrish) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/KillCompactionConfig.java:145]: are we emitting this metric?

File: server/src/test/java/org/apache/druid/server/coordinator/duty/KillAuditLogTest.java [hunk 3]
```diff
@@ -93,9 +94,16 @@ public void testConstructorFailIfInvalidPeriod()
         .withCoordinatorKillMaxSegments(10)
         .withCoordinatorKillIgnoreDurationToRetain(false)
         .build();
-    exception.expect(IllegalArgumentException.class);
-    exception.expectMessage("coordinator audit kill period must be >= druid.coordinator.period.metadataStoreManagementPeriod");
-    killAuditLog = new KillAuditLog(mockAuditManager, druidCoordinatorConfig);
+
+    final IllegalArgumentException exception = Assert.assertThrows(
+        IllegalArgumentException.class,
+        () -> killAuditLog = new KillAuditLog(mockAuditManager, druidCoordinatorConfig)
```
Comment (findingrish) [inline server/src/test/java/org/apache/druid/server/coordinator/duty/KillAuditLogTest.java:100]: why are we setting `killAuditLog`, couldn't this be simplified to 
```
Assert.assertThrows(
        IllegalArgumentException.class,
        () -> new KillAuditLog(mockAuditManager, druidCoordinatorConfig)
);
```

Comment (kfaraz) [reply server/src/test/java/org/apache/druid/server/coordinator/duty/KillAuditLogTest.java:100]: Yeah, I would prefer it to be so, but it shows up as an inspection warning if you don't assign the return value to some variable.

File: server/src/main/java/org/apache/druid/server/coordinator/duty/KillCompactionConfig.java (hunk 2 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/KillCompactionConfig.java:145]: Yes, all metrics collected in any coordinator duty are emitted in `DruidCoordinator.DutiesRunnable.emitStat`.

One of the motivations of this PR was to collect metrics everywhere and emit it from just the one place in `DruidCoordinator`. Other coordinator duties such as `RunRules` and `BalanceSegments` are already doing this.

**Review by findingrish — APPROVED** [review]
LGTM

**Review by AmatyaAvadhanula — APPROVED** [review]

**Comment by kfaraz** [comment]
Thanks a lot for the review, @findingrish , @AmatyaAvadhanula !

**Pull request merged**